omod/src/test/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProviderWebTest.java

```diff
@@ -71,43 +85,83 @@ public void getDiagnosticReportByWrongUuid_shouldReturn404() throws Exception {
 	
 	@Test
 	public void createDiagnosticReport_shouldCreateNewDiagnosticReport() throws Exception {
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
+		
 		DiagnosticReport diagnosticReport = new DiagnosticReport();
 		diagnosticReport.setId(DIAGNOSTIC_REPORT_UUID);
 		
-		MockHttpServletResponse response = post("/DiagnosticReport").accept(FhirMediaTypes.JSON).go();
+		when(service.saveDiagnosticReport(any(DiagnosticReport.class))).thenReturn(diagnosticReport);
+		
+		MockHttpServletResponse response = post("/DiagnosticReport").jsonContent(jsonDiagnosticReport)
+		        .accept(FhirMediaTypes.JSON).go();
 		
 		assertThat(response, isCreated());
 	}
 	
 	@Test
 	public void updateDiagnosticReport_shouldUpdateExistingDiagnosticReport() throws Exception {
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
+		
 		DiagnosticReport diagnosticReport = new DiagnosticReport();
 		diagnosticReport.setId(DIAGNOSTIC_REPORT_UUID);
 		
-		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).accept(FhirMediaTypes.JSON).go();
+		when(service.updateDiagnosticReport(anyString(), any(DiagnosticReport.class))).thenReturn(diagnosticReport);
+		
+		MockHttpServletResponse response = put("/DiagnosticReport/" + DIAGNOSTIC_REPORT_UUID)
+		        .jsonContent(jsonDiagnosticReport).accept(FhirMediaTypes.JSON).go();
 		
 		assertThat(response, isOk());
 	}
 	
 	@Test
 	public void updateDiagnosticReport_shouldErrorForIdMismatch() throws Exception {
-		DiagnosticReport diagnosticReport = new DiagnosticReport();
-		diagnosticReport.setId(DIAGNOSTIC_REPORT_UUID);
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
 		
-		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).accept(FhirMediaTypes.JSON).go();
+		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).jsonContent(jsonDiagnosticReport)
+		        .accept(FhirMediaTypes.JSON).go();
 		
-		// TODO: Fix so status is 400, not 200. See
-		assertThat(response, isOk());
+		assertThat(response, isBadRequest());
+		assertThat(response.getContentAsString(),
+		    containsStringIgnoringCase("body must contain an ID element which matches the request URL"));
 	}
 	
 	@Test
 	public void updateDiagnosticReport_shouldErrorForNoId() throws Exception {
```

Comment:
Not sure if this test is necessary, and if the error handling for updates in the DiagnosticReport and Task `ServiceImpl` are necessary either. The error handling gets handled by the HAPI libraries, and `when(service.updateDiatnosticReport())` is not needed for this and `ShouldErrorForIdMismatch()` tests. In other words, the error handling implementation for these to cases doesn't seem to have an effect on the handling of the actual request.

Reply:
Yeah, I mostly worry about using these tests to ensure that we've done our part of the mappings properly, i.e., when someone sends the expected message our code handles the response appropriately. Since this is purely a test of the HAPI library, I don't think we need it, provided it works with an appropriately formulated request.